integration/legacy_name_validation_test.go

```diff
@@ -0,0 +1,388 @@
+		labels.MustNewMatcher(labels.MatchEqual, "name", "ingester"),
+		labels.MustNewMatcher(labels.MatchEqual, "state", "ACTIVE"))))
+
+	client, err := e2emimir.NewClient(distributor.HTTPEndpoint(), querier.HTTPEndpoint(), "", "", userID)
+	require.NoError(t, err)
+
+	now := time.Now().Truncate(time.Millisecond).UnixMilli()
+
+	testCases := []struct {
+		name       string
+		rw1request *prompb.WriteRequest
+		rw2request *promRW2.Request
+	}{
+		{
+			name: "utf8 metric name",
+			rw1request: &prompb.WriteRequest{
+				Timeseries: []prompb.TimeSeries{
+					{
+						Labels:  []prompb.Label{{Name: "__name__", Value: "utf_metric😀C"}},
+						Samples: []prompb.Sample{{Timestamp: now, Value: 100}},
+					},
+				},
+				Metadata: []prompb.MetricMetadata{
+					{
+						MetricFamilyName: "utf_metric😀C_total",
+						Help:             "some helpC",
+						Unit:             "someunitC",
+						Type:             prompb.MetricMetadata_COUNTER,
+					},
+				},
+			},
+			rw2request: &promRW2.Request{
+				Timeseries: []promRW2.TimeSeries{
+					{
+						LabelsRefs: []uint32{0, 1},
+						Samples:    []promRW2.Sample{{Timestamp: now, Value: 100}},
+						Metadata: promRW2.Metadata{
+							Type:    promRW2.Metadata_METRIC_TYPE_COUNTER,
+							HelpRef: 2,
+							UnitRef: 3,
+						},
+					},
+				},
+				Symbols: []string{
+					"__name__", "utf_metric😀C_total",
+					"some helpC",
+					"someunitC",
+				},
+			},
+		},
+		{
+			name: "utf8 label name",
+			rw1request: &prompb.WriteRequest{
+				Timeseries: []prompb.TimeSeries{
+					{
+						Labels: []prompb.Label{
+							{Name: "__name__", Value: "legacy_metricC"},
+							{Name: "utf8_label😀", Value: "test"},
+						},
+						Samples: []prompb.Sample{{Timestamp: now, Value: 100}},
+					},
+				},
+			},
+			rw2request: &promRW2.Request{
+				Timeseries: []promRW2.TimeSeries{
+					{
+						LabelsRefs: []uint32{0, 1, 2, 3},
+						Samples:    []promRW2.Sample{{Timestamp: now, Value: 100}},
+					},
+				},
+				Symbols: []string{
+					"__name__", "legacy_metricC_total",
+					"utf8_label😀", "test",
+				},
+			},
+		},
+	}
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			t.Run("rw1", func(t *testing.T) {
+				res, err := client.PushRW1(tc.rw1request)
+				require.NoError(t, err)
+				require.Equal(t, res.StatusCode, http.StatusBadRequest)
+			})
+			t.Run("rw2", func(t *testing.T) {
+				res, err := client.PushRW2(tc.rw2request)
+				require.NoError(t, err)
+				require.Equal(t, res.StatusCode, http.StatusBadRequest)
+			})
+		})
+	}
+}
+
+// TestLegacyNameValidation_Read ensures that Mimir discards queries containing
+// invalid metric and label names.
+func TestLegacyNameValidation_Querier(t *testing.T) {
+	for _, queryEngine := range []string{"mimir", "prometheus"} {
+		t.Run(queryEngine, func(t *testing.T) {
+			testLegacyNameValidationQuerier(t, queryEngine)
+		})
+	}
+}
+
+func testLegacyNameValidationQuerier(t *testing.T, queryEngine string) {
```

Comment:
it might be more valuable to test this on the query-frontend. That's the component which will first receive the queries.